Openapi validation

Hypro999 opened the pull request:
This is an extension of PR #12681 and wraps up most of the work for issue #12521.
- The first commit cleans up the messiness for accounting for both `<var_name>` and `{var_name}` by standardizing on the `{var_name}` format.
- The second commit upgrades test which validates the request parameter types.
> TODO: This is lower priority than the main issue, but it'd be really cool to cross-validate the types declared in OpenAPI against the types declared in REQ (via the validator parameter). We may not be able to do this always, but 95% of the time it's check_int or check_str or check_dict, and we can certainly map those to the expected type.

However, instead of checking for the validators, I went ahead and did something even better. I analyse the function declaration itself (for more, please refer to the commit message).

Hypro999 commented:
On a sidenote, the test fails on Xenial with Python 3.5 but I don't know what the exact cause of it is (to be honest, I haven't investigated much yet either). It works flawlessly on Bionic with Python 3.6 though.

If I remember correctly, I read somewhere that the typing module had issues in Python 3.5 and that it was fixed in future versions.

Hypro999 commented:
@timabbott requesting review.

Hypro999 commented:
I'm going to force push to trigger the circleCI tests. Then we can look at the failure from Python3.5.

timabbott commented:
@Hypro999 thanks for working on this!  

The first commit looks great, but I'd really appreciate it if you split it up to be readable so I can verify whether the code changes to the test are correct -- currently there's code being changed and moved and deleted all mixed together.  E.g. a first commit that just migrates the existing documentation (maybe adding a `#nocoverage` if needed), and then a second commit that removes the code the `<parameter_name>` syntax, and then a third commit (or maybe one per function) that extract the functions.  

I'll post a few comments on the second commit as well.  I think the xenial failure is likely resolvable with some work; it was Trusty that didn't support this sort of thing at all.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -20,6 +23,14 @@
+    'array': list,
+    'Typing.List': list,
+    'NoneType': None,
+}
```
Just to be explicit, have you verified that if you edit something to have an incorrect type, this test fails?

Hypro999 replied:
Yup, I tested it out.

Hypro999 replied:
Also added a sample error message in a comment below this.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
 
+    def _highest_priority_type(self, types: List[type]) -> type:
+        priority = {list: 1, str: 2, int: 3, bool: 4}
+        tyiroirp = {1: list, 2: str, 3: int, 4: bool}
```
This sort of clever name is confusing; go with `type_to_priority` and `priority_to_type` for these.

Hypro999 replied:
```
======================================================================
FAIL: test_openapi_arguments (zerver.tests.test_openapi.OpenAPIArgumentsTest)
This end-to-end API documentation test compares the arguments
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/tests/test_openapi.py", line 457, in test_openapi_arguments
    self.check_argument_types(function, openapi_parameters)
  File "/srv/zulip/zerver/tests/test_openapi.py", line 350, in check_argument_types
    raise AssertionError(msg)
AssertionError: There seems to be a mismatch between the types for the request parameters documented in the OpenAPI documentation and the actual request parameters in the function decleration for update_message_flags.
=================================================================
Parameter          OpenAPI Type         Function Decleration Type
=================================================================
op                <class 'int'>         <class 'str'>


----------------------------------------------------------------------
```

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+decleration for {}.\n".format(function.__name__)
+            msg += '='*65 + '\n'
+            msg += "{:<10s}{:^30s}{:>10s}\n".format("Parameter", "OpenAPI Type", "Function Decleration Type")
+            msg += '='*65 + '\n'
```
Can you post an example of what this output looks like?

Hypro999 replied:
```
======================================================================
FAIL: test_openapi_arguments (zerver.tests.test_openapi.OpenAPIArgumentsTest)
This end-to-end API documentation test compares the arguments
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/tests/test_openapi.py", line 457, in test_openapi_arguments
    self.check_argument_types(function, openapi_parameters)
  File "/srv/zulip/zerver/tests/test_openapi.py", line 350, in check_argument_types
    raise AssertionError(msg)
AssertionError: There seems to be a mismatch between the types for the request parameters documented in the OpenAPI documentation and the actual request parameters in the function decleration for update_message_flags.
=================================================================
Parameter          OpenAPI Type         Function Decleration Type
=================================================================
op                <class 'int'>         <class 'str'>


----------------------------------------------------------------------
```

timabbott replied:
Great! :)

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+documented in the OpenAPI documentation and the actual request parameters in the function \
+decleration for {}.\n".format(function.__name__)
+            msg += '='*65 + '\n'
+            msg += "{:<10s}{:^30s}{:>10s}\n".format("Parameter", "OpenAPI Type", "Function Decleration Type")
```
"Declaration" is mispelled here and in the `msg` definition above.  You may also want to use ` """` heredoc to format the `msg` string better.

Hypro999 replied:
I fixed the spelling, but I'm not sure about how format the msg string better using multiline strings (heredocs). Are you referring to the new f-strings syntax (I'm guessing not since that's only available on Python 3.6+)?

timabbott replied:
msg = """
There seems to be a mistmatch between 
...
""".format(...)

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -309,21 +382,24 @@ def test_openapi_arguments(self) -> None:
-                # the types, too; `check_int` -> `int`, etc., and
-                # verify those too!
-                accepted_arguments = set(arguments_map[function])
+                accepted_arguments = set(arguments_map[function_name])
```
Let's split out a preparatory refactor commit that just renames `function` to `function_name`.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+                fp.add((vname, vtype))
+
+        diff = op - fp
+        if diff:  # nocoverage
```
If you extract this logic to a `render_openapi_type_exception` function, you can unit test that function, which might be a good idea given that the code is fairly complex.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+
+        fp = set()
+        while getattr(function, "__wrapped__", None):
+            function = function.__wrapped__  # type: ignore # mypy doesn't recognize the condition
```
You could just do `assert function is not None` here to avoid the `type: ignore`; that'd be clearer.

Hypro999 replied:
@timabbott, actually, the reason for why I use `#type: ignore` is that mypy complains:
> error: "Callable[..., Any]" has no attribute "__wrapped__".

Since it does not know that the we are doing the `getattr` in the line above.

timabbott replied:
Oh, well, you could probably fix that by writing e.g.

`function = getattr(function, "__wrapped__", None)
assert function is not None`

Hypro999 replied:
This works, and is a pretty convenient hack. I've added it.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+            defval = defval.default
+            if defval.__class__ == REQ:
+                vtype = self._get_standardized_argument_type(function.__annotations__[vname])
+                vname = defval.post_var_name  # type: ignore # it's a REQ; account for whence
```
If you do a `type: REQ` on the previous line, I think you won't need this ignore.

Hypro999 replied:
It doesn't really work...
>mypy      | zerver/tests/test_openapi.py:363: error: Invalid type "zerver.lib.request.REQ"
mypy      | zerver/tests/test_openapi.py:364: error: REQ? has no attribute "__class__"
mypy      | zerver/tests/test_openapi.py:366: error: REQ? has no attribute "post_var_name"

timabbott replied:
For fairly deep mypy internal reasons (basically making type-checking inside our view functions work nicely), `REQ` has a technically incorrect type stub, defined in `zerver/lib/request.pyi`.  So it's possible my previous statement was wrong and there's no way to do this properly without ignores.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+                  element in openapi_parameters])
+
+        fp = set()
+        while getattr(function, "__wrapped__", None):
```
Maybe add a comment just before this line explaining 

"Iterate through the decorators to find the original function so we can parse its arguments."

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+        op = set([(element["name"], VARMAP[element["schema"]["type"]]) for
+                  element in openapi_parameters])
+
+        fp = set()
```
Give this a better name; I'm not sure what `fp` stands for.  In general we don't use abbreviations in our code because it often results in code that isn't easy for people other than the author to understand (and even for the author they may forget 6 months later)

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+    def check_argument_types(self, function: Callable[..., HttpResponse],
+                             openapi_parameters: List[Dict[str, Any]]) -> None:
+        op = set([(element["name"], VARMAP[element["schema"]["type"]]) for
+                  element in openapi_parameters])
```
Similarly with `op` -- I guess this is `openapi_parameter_tuples`? or something?

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -309,21 +382,24 @@ def test_openapi_arguments(self) -> None:
+                    function_name, tags = value
+
+                lookup_parts = function_name.split('.')
+                module = __import__('.'.join(lookup_parts[:-1]), {}, {}, [''])
```
Does this eliminate the need to have those imports at the top of the class?  If so, might make sense to extract a prep commit that replaces those imports with this.

Hypro999 replied:
@timabbott: Ah, no, this does not. In fact this import isn't even an import (I think), what we're doing is just retrieving the function from memory.

So my logic here was that by importing urls.py, since we've already imported every view function (which is what triggers all of the has_request_variables decorators), all of the functions are already loaded into runtime memory. Thus with this "import" we're just getting a handle to that function.

Hypro999 replied:
I've explained more in the (rather lengthy) commit message.

timabbott replied:
You're correct that if we've already imported a module, this will just look it up and not import it again.  However, if we hadn't, this line will import it.  Which is why I think this should be able to allow us to remove the views imports for views not always imported during Django setup, above.

Hypro999 replied:
You're correct, we can remove those lines. Good observation! Thank you for the suggestion!

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+        return self._get_standardized_argument_type(t.__args__[0])
+
+    def check_argument_types(self, function: Callable[..., HttpResponse],
+                             openapi_parameters: List[Dict[str, Any]]) -> None:
```
Let's add a docstring here, explaining how this check works:

"We construct for both the OpenAPI data and the function's definition a set of tuples of the form
`(var_name, type)` and then compare those sets to see if the OpenAPI data defines a different
type than that actually accepted by the function."

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -276,6 +287,68 @@ def check_for_non_existant_openapi_endpoints(self) -> None:
+                val = v
+        return tyiroirp.get(val, types[0])
+
+    def _get_standardized_argument_type(self, t: Any) -> type:
```
This function could also use a docstring:

"Given a type from the `typing` module such as `List[str]` or `Union[str, int]`, convert it into
a corresponding Python type.  Unions are mapped to a canonical choice among the options."
"

timabbott commented:
I think the traceback in Xenial CircleCI for `typing.Union[typing.List<~T>[typing.Dict<~KT, ~VT>[str, typing.Any]], NoneType]` suggests just that there's a bug in the Python 3.5 logic for converting argument types to the canonical form.

Overall, the type-checking stuff is awesome.  I think there's probably a further iteration we could do by further comparing the types to the `check_list` type functions, basically just helping avoid mistakes in writing view functions, but we should definitely think of that as a follow-up since it's not relevant to getting the main value of this project.

timabbott commented:
I merged the first commit after this change, identified with some grepping:

```
diff --git a/templates/zerver/api/get-presence.md b/templates/zerver/api/get-presence.md
index 37fc1f3856..09deb1a4b9 100644
--- a/templates/zerver/api/get-presence.md
+++ b/templates/zerver/api/get-presence.md
@@ -8,7 +8,7 @@ Zulip clients like mobile/desktop apps will want to use the main
 presence endpoint, which returns data for all active users in the
 organization, instead.
 
-`GET {{ api_url }}/v1/users/<email>/presence`
+`GET {{ api_url }}/v1/users/{email}/presence`
 
 See
 [Zulip's developer documentation](https://zulip.readthedocs.io/en/latest/subsystems/presence.html)
@@ -24,7 +24,7 @@ for details on the data model for presence in Zulip.
 {tab|curl}
 
 ``` curl
-curl -X GET {{ api_url }}/v1/users/<email>/presence \
+curl -X GET {{ api_url }}/v1/users/{email}/presence \
     -u BOT_EMAIL_ADDRESS:BOT_API_KEY
 ```
 
```

timabbott commented:
I merged all but the last commit.  It's really exciting to see this coming together.

@eeshangarg @YagoGG @sameerchoubey FYI, you may enjoy reading `OpenAPIArgumentsTest` -- it basically validates our API parameters documentation against the actual API checking for inconsistencies in the list of parameters (and the remaining commit in this PR checks the types too!).

timabbott commented:
@Hypro999 is this stuck on figuring out the test failure on Xenial?  I think I know what's going on:

That endpoint has `converter=extract_recipients`, which generally means the type passed by the user in the API doesn't necessarily match the type returned by the function.  We should probably just exclude those from this type-checking tooling; it's only with `validator` that we expect that to be true.

timabbott commented:
I pushed some updates to this branch that fix an issue I could reproduce locally on Bionic.

Hypro999 commented:
@timabbott, I fixed the Python 3.5 issue. The main issue was that `typing.Union` was implemented in a different way in Python 3.5 than in future versions. I also squashed the commit you made to make sure that each commit is stable (thanks for the help there!). This is finally ready to be merged :grin:.

timabbott commented:
This is great, merged, thanks @Hypro999!  I added a comment explaining the TODO involving `converter` elements; that may become relevant as we document more endpoints.

I think there is some room for further checking here, but I think it's probably better for you to focus on the other things I mentioned in that OpenAPI documentation thread to simplify the steps required to add a new endpoint.  I think in particular the proposal of making management of `TEST_FUNCTIONS` be a decorator would make a huge difference in how error-prone the current setup is.
